### PR TITLE
Finished stuffing 2

### PR DESCRIPTION
**Pull request by ekr**
Work in progress.

### CODE REVIEW

**Comment by ekr** [comment]
I'm a little iffy on this "multiple PSKs in PSK binder" thing. @mt @davidben would be interested in your view. Note that it's now very easy to change to only one PSK allowed, because we just make this singular

**Comment by ekr** [comment]
Three notes here:
1. Karthik suggests combining PSK and psk_binder. I agree, so I'll do that in a bit.
2. We may want to have a separate ticket_age_add and edi_allowed extension in NST because that will let us do freshness in 0.5RTT
3. We should compute the binder separately for resumption and external PSK. A different label will do it.

File: draft-ietf-tls-tls13.md [hunk 1]
```diff
@@ -2229,43 +2265,56 @@ the client and servers MUST supply "key_share" values as described
...
+each PSK offered in the "pre_shared_keys" extension and in the same
+order. The "psk_binder" extension MUST be sent if the client offers a
+"pre_shared_key" extension and MUST NOT be sent otherwise.
 
-In order to accept PSK key establishment, the server sends an empty
-"pre_shared_key" extension.
+Each binding_data is computed the same way as the Finished.verify data
+value, using the PSK as the base key and the portion of the handshake
+transcript up to but not including the PskBinder extension as the
+Handshake Context. I.e.,
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: "Handshake Context, i.e., "

File: draft-ietf-tls-tls13.md [hunk 2]
```diff
@@ -2305,7 +2354,8 @@ If it is not, the server SHOULD proceed with the handshake but reject
 
 The parameters for the 0-RTT data (symmetric cipher suite,
 ALPN, etc.) are the same as those which were negotiated in the connection
-which established the PSK.
+which established the PSK. The PSK used to encrypt the early data
+MUST be the first PSK listed in the client's "pre_shared_key"
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Missing period

File: draft-ietf-tls-tls13.md [hunk 3]
```diff
@@ -3980,16 +3983,20 @@ TLS-compliant application MUST implement the following TLS extensions:
...
 
   * "supported_versions" is REQUIRED for all ClientHello messages.
-  * "signature_algorithms" is REQUIRED for certificate authenticated cipher suites.
-  * "supported_groups" and "key_share" are REQUIRED for DHE or ECDHE cipher suites.
-  * "pre_shared_key" is REQUIRED for PSK cipher suites.
+  * "signature_algorithms" is REQUIRED for certificate authentication.
+  * "supported_groups" and "key_share" are REQUIRED for DHE or ECDHE key
+    establishment.
+  * "pre_shared_key" is REQUIRED for PSK key establishment.
+  * "psk_key_exchange_modes" is required when PSKs are offered.
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: +psk_binder

File: draft-ietf-tls-tls13.md [hunk 4]
```diff
@@ -3980,16 +3983,20 @@ TLS-compliant application MUST implement the following TLS extensions:
   * Negotiated Groups ("supported_groups"; {{negotiated-groups}})
   * Key Share ("key_share"; {{key-share}})
   * Pre-Shared Key ("pre_shared_key"; {{pre-shared-key-extension}})
+  * Pre-Shared Key Exchange Modes ("psk_key_exchange_modes";
+    {{pre-shared-key-exchange-modes}})
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: + psk_binder

Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Though I have to ask whether you don't need to revisit this section.  It's possible not to support PSK.

File: draft-ietf-tls-tls13.md [hunk 5]
```diff
@@ -3980,16 +3983,20 @@ TLS-compliant application MUST implement the following TLS extensions:
...
   * Server Name Indication ("server_name"; Section 3 of {{RFC6066}})
   * Cookie ("cookie"; {{cookie}})
 
 All implementations MUST send and use these extensions when offering
 applicable cipher suites:
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Note that this isn't based on cipher suites any more, it's based on key exchange modes.  I would rearrange this section and say:

If you support Diffie-Hellman key exchange, you need key_share, supported_groups, and signature_algorithms.  If you support PSK, you need pre_shared_key, psk_key_exchange_modes and psk_binder.  It's also possible to support a a combination of both that excludes support for signature_algorithms if you only support DHE+PSK with PSK authentication...

**Pull request closed without merging**